### PR TITLE
8342936: Enhance java.io.IO with parameter-less println() and readln()

### PR DESCRIPTION
**Pull request by lahodaj**
This PR is simply adding parameter-less `java.io.IO.{println(),readln()}`, with the (hopefully) obvious semantics, plus the corresponding wiring to make those work. This may become part of JEP 495:
https://openjdk.org/jeps/495

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8342937](https://bugs.openjdk.org/browse/JDK-8342937) to be approved
- [x] Change requires a JEP request to be targeted

### Issues
 * [JDK-8342936](https://bugs.openjdk.org/browse/JDK-8342936): Enhance java.io.IO with parameter-less println() and readln() (**Sub-task** - P4)
 * [JDK-8335984](https://bugs.openjdk.org/browse/JDK-8335984): JEP 495: Simple Source Files and Instance Main Methods (Fourth Preview) (**JEP**)
 * [JDK-8342937](https://bugs.openjdk.org/browse/JDK-8342937): Enhance java.io.IO with parameter-less println() and readln() (**CSR**)

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**) Review applies to [78bf15bf](https://git.openjdk.org/jdk/pull/21693/files/78bf15bf1ad03a535b04727b1a982c7b440410e7)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21693/head:pull/21693` \
`$ git checkout pull/21693`

Update a local copy of the PR: \
`$ git checkout pull/21693` \
`$ git pull https://git.openjdk.org/jdk.git pull/21693/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21693`

View PR using the GUI difftool: \
`$ git pr show -t 21693`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21693.diff">https://git.openjdk.org/jdk/pull/21693.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21693#issuecomment-2436189851)
</details>

### CODE REVIEW

**Review by asotona — APPROVED** [review]
Looks good to me.

**Review by naotoj — APPROVED** [review]
+1

**Comment by jaikiran** [comment]
Hello Jan, the GitHub actions job failures look related:

```
/Users/runner/work/jdk/jdk/test/langtools/jdk/jshell/ConsoleTest.java:229: error: ThrowingJShellConsole is not abstract and does not override abstract method readLine() in JShellConsole
    private static class ThrowingJShellConsole implements JShellConsole {
                   ^
/Users/runner/work/jdk/jdk/test/langtools/jdk/jshell/ConsoleTest.java:199: error: <anonymous ConsoleTest$11> is not abstract and does not override abstract method readLine() in JShellConsole
        super.setUp(bc.andThen(b -> b.console(new JShellConsole() {
                                                                  ^
2 errors
result: Failed. Compilation failed: Compilation failed
```

**Comment by lahodaj** [comment]
> Hello Jan, the GitHub actions job failures look related:

Yes, sorry for that. Should be fixed now.

File: src/java.base/share/classes/java/io/Console.java [hunk 1]
```diff
@@ -172,6 +172,19 @@ public Console println(Object obj) {
...
+    /**
+     * Terminates the current line in this console's output stream using
+     * {@link System#lineSeparator()} and then flushes the console.
+     *
+     * @return  This console
+     *
+     * @since 24
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.IMPLICIT_CLASSES)
+    public Console println() {
```
Comment (jaikiran) [inline src/java.base/share/classes/java/io/Console.java:184]: Hello Jan, with the introduction of the simplified `java.io.IO` class, some of whose APIs are wrappers around the `java.io.Console` class APIs, it's understandable that we will start seeing new methods being introduced on the `Console` class.

For this specific new `println()` method on the `Console` class, do you think this method will/should be introduced irrespective of the `IO` class usage? In other words, considering a theoretic case where there may be a situation to abandon the implicit classes preview feature or do it in a different way, would we still want this `println()` method to stay on the `Console`? If yes, then perhaps we should drop the `@PreviewFeature` annotation from this new method and introduce it as a regular new API? Same question for the new `readln()` method on the `Console`.

File: src/java.base/share/classes/java/io/Console.java [hunk 2]
```diff
@@ -214,6 +227,24 @@ public String readln(String prompt) {
...
+     * Reads a single line of text from this console.
+     *
+     * @throws IOError
+     *         If an I/O error occurs.
+     *
+     * @return  A string containing the line read from the console, not
+     *          including any line-termination characters, or {@code null}
+     *          if an end of stream has been reached without having read
+     *          any characters.
+     *
```
Comment (jaikiran) [inline src/java.base/share/classes/java/io/Console.java:240]: Should we specify (both here and in the `IO.readln()`) what is returned if only the line-terminator character is read from the console? A brief look at the implementation seems to suggest that we return an empty `String`.

Comment (Evemose) [reply src/java.base/share/classes/java/io/Console.java:240]: I'm not an expert in this field, but It is pretty common to exclude line terminator from readln string in different languages. Therefore, It should be intuitive that if nothing was present in line, the result of method will be string with noting, i.e. empty string. The method could benefit from some sort of semantic annotation like NonNull, but I guess modern IDEs do pretty good job on inferring such things anyway.

PS: Also I feel like for person that never wrote code in their life "line terminator" would sound like some sort of error, maybe the better wording should be used in IO specifically

File: src/java.base/share/classes/java/io/Console.java (hunk 1 above)
Comment (lahodaj) [reply src/java.base/share/classes/java/io/Console.java:184]: I don't think this PR is the right place to add the methods permanently to `Console`. I am not objecting to that, but it seems the if those would be added independently of `java.io.IO`, it needs a separate discussion and decision. And I believe that, if needed, that should be done outside of and independently on the framework of `java.io.IO` and JEP 495, as both of these are preview features.

File: src/java.base/share/classes/java/io/Console.java (hunk 2 above)
Comment (lahodaj) [reply src/java.base/share/classes/java/io/Console.java:240]: I also find the text to be fairly clear.

**Review by asotona — APPROVED** [review]

File: src/java.base/share/classes/java/io/Console.java (hunk 1 above)
Comment (jaikiran) [reply src/java.base/share/classes/java/io/Console.java:184]: Hello Jan, I agree that if we are going to introduce the parameter-less `println()` and `readln()` APIs on `Console` class independently of the preview feature, then it should be done in a separate PR and with a separate discussion. I was going to suggest that we should probably do that, but I notice that the existing `readln` and `println` methods which take a parameter are already marked preview feature methods in Java 23 https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/io/Console.html#preview-println(java.lang.Object). Given that, I believe what you have here in this PR - marking these new parameter-less methods on `Console` as preview feature, is the right thing to do.

**Review by jaikiran — APPROVED** [review]

**Comment by lahodaj** [comment]
> My question (please don't bash me, I am new): Any chance we can set the encoding of console? I tested on Java 23 (Windows) and the encoding is `windows-1252` so it will not display unicode characters correctly.

I believe that after https://github.com/openjdk/jdk/pull/21569, the `Console` encoding used by the JDK can be overwritten with `stdout.encoding`.

**Comment by naotoj** [comment]
> > My question (please don't bash me, I am new): Any chance we can set the encoding of console? I tested on Java 23 (Windows) and the encoding is `windows-1252` so it will not display unicode characters correctly.
> 
> I believe that after #21569, the `Console` encoding used by the JDK can be overwritten with `stdout.encoding`.

Yes, that is in place. Will be available in JDK24: https://bugs.openjdk.org/browse/JDK-8342208

**Pull request closed without merging**